prescient/plugins/tracking.py

```diff
@@ -0,0 +1,1619 @@
+                    g_stat_list.append(round(pyo.value(stats_var_dict[s][g,t]),2))
+
+        return
+
+    def _record_results(self, m, market_signals, date = None, hour = None, **kwargs):
+
+        df_list = []
+        for generator in m.UNITS:
+            for t in m.HOUR:
+
+                result_dict = {}
+                result_dict['Generator'] = generator
+                result_dict['Date'] = date
+                result_dict['Hour'] = hour
+
+                # simulation inputs
+                result_dict['Horizon [hr]'] = int(t)
+                result_dict['Market Dispatch [MW]'] = float(round(market_signals[generator][t],2))
+                result_dict['Market Commitment [bin]'] = int(result_dict['Market Dispatch [MW]'] > 0)
+
+                # model vars
+                result_dict['Thermal Power Generated [MW]'] = float(round(pyo.value(m.P_T[generator,t]),2))
+
+                result_dict['On/off [bin]'] = int(round(pyo.value(m.on_off[generator,t])))
+                result_dict['Start Up [bin]'] = int(round(pyo.value(m.start_up[generator,t])))
+                result_dict['Shut Down [bin]'] = int(round(pyo.value(m.shut_dw[generator,t])))
+
+                result_dict['Production Cost [$]'] = float(round(pyo.value(m.prod_cost_approx[generator,t]),2))
+                result_dict['Start-up Cost [$]'] = float(round(pyo.value(m.start_up_cost_expr[generator,t]),2))
+                result_dict['Total Cost [$]'] = float(round(pyo.value(m.tot_cost[generator,t]),2))
+
+                # calculate mileage
+                if t == 0:
+                    result_dict['Mileage [MW]'] = float(round(abs(pyo.value(m.P_T[generator,t] - m.pre_P_T[generator])),2))
+                else:
+                    result_dict['Mileage [MW]'] = float(round(abs(pyo.value(m.P_T[generator,t] - m.P_T[generator,t-1])),2))
+
+                for key in kwargs:
+                    result_dict[key] = kwargs[key]
+
+                result_df = pd.DataFrame.from_dict(result_dict,orient = 'index')
+                df_list.append(result_df.T)
+
+        # save the result to object property
+        # wait to be written when simulation ends
+        self.result_list.append(pd.concat(df_list))
+
+        return
+
+    def get_last_delivered_power(self,generator):
+        return pyo.value(self.model.pre_P_T[generator])
+
+    def write_results(self,path):
+
+        print("")
+        print('Saving tracking results to disk...')
+        pd.concat(self.result_list).to_csv(os.path.join(path,'tracking_detail.csv'), \
+                                           index = False)
+        return
+
+class DAM_hybrid_tracking:
+
+    def __init__(self,rts_gmlc_data_dir, generators = None):
+        self.n_scenario = n_scenario
+        self.model_data = self.assemble_model_data(generator_names = generators, \
+                                                   rts_gmlc_data_dir = rts_gmlc_data_dir)
+
+    def assemble_model_data(self,generator_names, \
```

Comment:
How is this different than the other `assemble_model_data` methods? Is there an opportunity to abstract this to really shrink the code base?